### PR TITLE
Adds necessary shinytest2 helpers for automated tests

### PR DESCRIPTION
**Pull request by averissimo**
# Pull Request

Part of #712

#### Changes description

- Allow `TealAppDriver` to be used in tests
  - R6 class is not yet exported from `{teal}` and needs to be used here
- Adds CDISC simple data helper to use in e2e tests
- Adds `skip_if_too_deep` function

### CODE REVIEW

File: tests/testthat/helper-testing-depth.R [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#' @param depth `numeric` the depth of the testing evaluation,
+#'   has opposite interpretation to environment variable `TESTING_DEPTH`.
+#'   So e.g. `0` means run it always and `5` means a heavy test which should be run rarely.
+#'   If the `depth` argument is larger than `TESTING_DEPTH` then the test is skipped.
+#' @importFrom testthat skip
+#' @return `NULL` or invoke an error produced by `testthat::skip`
+#' @note By default `TESTING_DEPTH` is equal to 3 if there is no environment variable for it.
+#' By default `depth` argument lower or equal to 3 will not be skipped because by default `TESTING_DEPTH`
+#' is equal to 3. To skip <= 3 depth tests then the environment variable has to be lower than 3 respectively.
+skip_if_too_deep <- function(depth) { # nolintr
```
Comment (averissimo) [inline tests/testthat/helper-testing-depth.R:43]: @pawelru @vedhav Is there any requirement/edge case for placing these functions on `testthat/setup-*` instead of `helper-*`?

BLUF: The skip_if_too_deep function itself isn't available on `devtools::load_all` and has to be skipped when running tests manually.

It's a small grip, but I think all code on tests should be "runnable". Moreover, when reading the documentation it seems to support moving this function to `helpers-*.R`

https://testthat.r-lib.org/articles/special-files.html?q=options#helper-files

image


image

Comment (pawelru) [reply tests/testthat/helper-testing-depth.R:43]: I agree with you. This feels to me to be more of a helper- instead of setup- type of content.

File: tests/testthat/helper-TealAppDriver.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  datasets <- match.arg(datasets, several.ok = TRUE)
+  data <- within(teal.data::teal_data(), {
+    require(nestcolor)
+    ADSL <- teal.modules.general::rADSL
+    ADRS <- teal.modules.general::rADRS
+    ADTTE <- teal.modules.general::rADTTE
+  })
+
+  teal.data::datanames(data) <- datasets
+  data
```
Comment (kartikeyakirar) [inline tests/testthat/helper-TealAppDriver.R:25]:
```suggestion
  datasets <- match.arg(datasets, several.ok = TRUE)
  within(teal.data::cdisc_data(join_keys = teal.data::default_cdisc_join_keys[datasets]), {
    require(nestcolor)
    ADSL <- teal.modules.general::rADSL
    ADRS <- teal.modules.general::rADRS
    ADTTE <- teal.modules.general::rADTTE
  })
```

Comment (averissimo) [reply tests/testthat/helper-TealAppDriver.R:25]: Good catch with the join_keys :+1:

Given that `cdisc_data` does nothing different than `teal_data`  if no named arguments are given, what do you think about keep using `teal_data()`?

```suggestion
  datasets <- match.arg(datasets, several.ok = TRUE)
  data <- within(
    teal.data::teal_data(join_keys = teal.data::default_cdisc_join_keys[datasets]), 
    {
      require(nestcolor)
      ADSL <- teal.modules.general::rADSL
      ADRS <- teal.modules.general::rADRS
      ADTTE <- teal.modules.general::rADTTE
    }
  )
```

Comment (kartikeyakirar) [reply tests/testthat/helper-TealAppDriver.R:25]: then you have to assign datanames .  I am ok with either of them.

**Review by kartikeyakirar — APPROVED** [review]
LGTM 🚀
